### PR TITLE
Add framework for extensions in API spec

### PR DESCRIPTION
**Pull request by oddhack**
Now integrates all khr extensions. Build with all extensions via 

$ python3 makeSpec -clean -spec khr apihtml chtml (other 'make' arguments if needed)

or build with no extensions included via

$ python3 makeSpec -clean -spec core <targets>

Note: 'makeAll' script has been replaced by 'makeSpec' which has slightly different syntax.

### CODE REVIEW

**Comment by bashbaug** [comment]
This is looking really nice!  A few comments from a brief review:

1. It looks like some of the heading levels are slightly off for the extensions so the table of contents looks really bad.  I think the extension name should be at the current level, but all other heading should be increased one level.
2. The extension APIs don't seem to be linking at all, see for example `clCreateCommandQueueWithPropertiesKHR`.
3. Most of the extension types don't seem to be linking at all either, though for some of these it's not clear what they would be linking to.  The one exception I found is the new structure for `cl_device_integer_dot_product_acceleration_properties_khr`, which is linking nicely.
4. The extension enums I tried are links, but they don't seem to be linking to anything.  Not sure what's going on here.
5. Editorial comment: some of the column widths seem to be "off" now, especially the `clGetDeviceInfo` table, where the leftmost column is very wide now.  I suspect this is because it's sizing the column to fit the very long enum `CL_DEVICE_INTEGER_DOT_PRODUCT_ACCELERATION_PROPERTIES_4x8BIT_PACKED_KHR` text, which doesn't have `<wbr>` tags like the asciidoctor attributes do?  One "solution" to fix this would be to omit the duplicated enum name from the text "CL_DEVICE_INTEGER_DOT_PRODUCT_ACCELERATION_PROPERTIES_4x8BIT_PACKED_KHR requires cl_khr_integer_dot_product", and just say something like "Requires cl_khr_integer_dot_product", which would be consistent with our other version text like "[Missing before](https://registry.khronos.org/OpenCL/specs/3.0-unified/html/OpenCL_API.html#unified-spec) version 1.1."
6. It'd be nice if there were some indicator tying an extension structure (like cl_device_integer_dot_product_acceleration_properties_khr) to the extension name, similar to the extension enums.

**Comment by oddhack** [comment]
> This is looking really nice! A few comments from a brief review:
> 
> 1. It looks like some of the heading levels are slightly off for the extensions so the table of contents looks really bad.  I think the extension name should be at the current level, but all other heading should be increased one level.

Yes. I'm holding off on adjusting that sort of stuff in order to do it in one pass. Also there will be some increased autogeneration in the extension appendices before I'm done, for the interface summaries - the existing markup is largely a placeholder from the separate documents.

> 2. The extension APIs don't seem to be linking at all, see for example `clCreateCommandQueueWithPropertiesKHR`.
> 3. Most of the extension types don't seem to be linking at all either, though for some of these it's not clear what they would be linking to.  The one exception I found is the new structure for `cl_device_integer_dot_product_acceleration_properties_khr`, which is linking nicely.
> 4. The extension enums I tried are links, but they don't seem to be linking to anything.  Not sure what's going on here.

All the extension appendices in the API spec have been done now, but the actual integration of changes / additions to the spec body remains TBD for a lot of them, which is what I'm focusing on now. Also, I'm not entirely clear on the conventions for the various asciidoctor attributes in the dictionaries so would appreciate some thoughts on (for example) these attributes extracted from the generated api-dictionary.asciidoc:

* cl_device_integer_dot_product_capabilities_khr_TYPE - formatted text with breaks
* cl_mutable_dispatch_exec_info_khr_TYPE_label - formatted text with breaks
* cl_mutable_dispatch_exec_info_khr_TYPE - xref
* clGetPlatformIDs_label - formatted text (perhaps with breaks?)
* clGetPlatformIDs - xref
* CL_CHAR_BIT_label - formatted text with breaks
* CL_CHAR_BIT - xref
* CL_CHAR_BIT_anchor - anchor

This is somewhat inconsistent:

* why are the type names suffixed with '_TYPE' but the command / enum names are not?
* cl_device_integer_dot_product_capabilities_khr (enumerated type) and cl_mutable_dispatch_exec_info_khr (struct) are treated differently even though they're both types:
  * the former gets only the formatted text attribute
  * the latter gets both formatted text and xref attributes

ISTM it would be more consistent to have

* {API_label} -> text (always)
* {API} -> xref (always),
* {API_anchor} -> anchor (only needed for enum names, I think)

and no _TYPE suffixes at all (except for API names that actually contain _TYPE, as some enums do). That would cause some markup changes, though most of the 400-odd uses are just basic scalar (cl_uint etc.) names.

> 5. Editorial comment: some of the column widths seem to be "off" now, especially the `clGetDeviceInfo` table, where the leftmost column is very wide now.  I suspect this is because it's sizing the column to fit the very long enum `CL_DEVICE_INTEGER_DOT_PRODUCT_ACCELERATION_PROPERTIES_4x8BIT_PACKED_KHR` text, which doesn't have `<wbr>` tags like the asciidoctor attributes do?  One "solution" to fix this would be to omit the duplicated enum name from the text "CL_DEVICE_INTEGER_DOT_PRODUCT_ACCELERATION_PROPERTIES_4x8BIT_PACKED_KHR requires cl_khr_integer_dot_product", and just say something like "Requires cl_khr_integer_dot_product", which would be consistent with our other version text like "[Missing before](https://registry.khronos.org/OpenCL/specs/3.0-unified/html/OpenCL_API.html#unified-spec) version 1.1."

Will do.

> 6. It'd be nice if there were some indicator tying an extension structure (like cl_device_integer_dot_product_acceleration_properties_khr) to the extension name, similar to the extension enums.

Also will do. Thanks for the feedback!

**Comment by bashbaug** [comment]
> This is somewhat inconsistent:
> 
> * why are the type names suffixed with '_TYPE' but the command / enum names are not?
> * cl_device_integer_dot_product_capabilities_khr (enumerated type) and cl_mutable_dispatch_exec_info_khr (struct) are treated differently even though they're both types:
>   
>   * the former gets only the formatted text attribute
>   * the latter gets both formatted text and xref attributes

Yes, I agree this is a little confusing.

The reason the type names are suffixed with `_TYPE` is because asciidoctor attributes are not case-sensitive so we cannot otherwise differentiate between `cl_float` (the type) and `CL_FLOAT` (the enum).  More info here: https://github.com/KhronosGroup/OpenCL-Docs/pull/444

We could have added a suffix to everything but that hasn't been needed so far.

IIRC the struct types get handled differently because the struct description gets generated from the XML and included in the spec so we have something to link to.  The non-struct types do not, so we do not include a link for them.  If we added a description for the other types (say, in an appendix) then we could link to them also, but we do not do this currently, and I'm not sure how much value it would add.

**Comment by oddhack** [comment]
@bashbaug do you have much Travis expertise? I needed to tweak the dependency-generating actions in the Makefile to deal with a generated file include, and this resulted in a truly obscure error that only happens when running under Travis CI, not locally, and only in one particular way of getting at the Makefile construct in question, through variable assignment rather than use directly in a gmake action. I've pushed the question off to https://stackoverflow.com/questions/77786801/how-could-running-in-travis-ci-affect-behavior-of-shell-in-a-makefile after a very annoying couple of hours doing edit-commit-wait for CI errors loops and getting nowhere, and am about at the point of just lifting the entire shell pipeline involved into a separate shell script.

**Comment by oddhack** [comment]
OK, went ahead and punted it to a scripts/find_adoc_deps scripts to make progress.

@bashbaug what would you think of migrating from Travis to GH Actions? We use it for Vulkan CI with a custom Docker image we rolled that has all the spec tools pre-installed, and advise people to run the Docker image rather than try to provide installation instructions on lots of different platforms (though they still can, by replicating the Dockerfile component installation, of course).

**Review by EwanC — COMMENTED** [review]
Left some minor command-buffer related comments from looking at the diff.

When I try to build this branch with `make html` I see the error `', needed by 'out/html/OpenCL_ICD_Installation.html'.  Stop.` And with the `make html` target none of the appendices seem to be included in `OpenCL_API.html`, so i'm not sure how to review the rendered output.

File: api/cl_khr_command_buffer_multi_device.asciidoc
Comment (oddhack) [reply api/cl_khr_command_buffer_multi_device.asciidoc:64]: I am hoping to replace all these handcoded interface sections with autogenerated content from the XML, as Vulkan already does. The one thing holding me back was that there are additional annotations, like this, beyond just the name of the API element itself. If the annotations aren't required then I can do that quickly, but it affects more than just this one extension appendix so at least @bashbaug and perhaps the entire WG should give feedback on that change.

**Comment by oddhack** [comment]
> Left some minor command-buffer related comments from looking at the diff.

Thanks!
 
> When I try to build this branch with `make html` I see the error `', needed by 'out/html/OpenCL_ICD_Installation.html'. Stop.` And with the `make html` target none of the appendices seem to be included in `OpenCL_API.html`, so i'm not sure how to review the rendered output.

Are you sure it's up to date with github adoc branch? I don't see this. However, building with just 'make' will not include any extensions in the specs at all. You need to use './makeAll html', which passes additional arguments to include all extensions. N.b. I will be replacing that script with a more flexible version from Vulkan soon.

File: api/cl_khr_command_buffer_mutable_dispatch.asciidoc
Comment (oddhack) [reply api/cl_khr_command_buffer_mutable_dispatch.asciidoc:115]: This should be fixed now.

**Comment by EwanC** [comment]
> Are you sure it's up to date with github adoc branch? I don't see this. However, building with just 'make' will not include any extensions in the specs at all. You need to use './makeAll html', which passes additional arguments to include all extensions. N.b. I will be replacing that script with a more flexible version from Vulkan soon.

Apologies, yup `./makeAll` works. You documented that in the PR description and I still missed it. Will review that local rendered HTML.

**Comment by oddhack** [comment]
Also, it could be the gmake version you're running? I had a really weird bug recently which I thought was Travis-related, but turned out to be because Travis was running an older gmake version than my 4.3, with a very subtle behavior difference.

**Comment by bashbaug** [comment]
Hi @oddhack, we discussed these changes in our teleconference on Tuesday and it seems like it would be easiest and less error-prone to merge these changes sooner rather than later.  This will also give us more time to make any necessary refinements before we need to publish the next version of the spec.

Since this PR is still makred as a "draft" are there other change you are hoping to make before merging, or are we OK to merge in the current state?

Are there any specific items you would like us to review before merging?

Thanks!

**Comment by oddhack** [comment]
> Since this PR is still makred as a "draft" are there other change you are hoping to make before merging, or are we OK to merge in the current state?
> 
> Are there any specific items you would like us to review before merging?

I'd like to bring it up to date with recent changes in some of the mutable extensions, which have been coming pretty fast. Are there likely to be any changes in those ext specs this week?

Once that's done, it would be best to migrate all development on published khr extensions over to the API and C language specs, and actually remove the separate extension documents from the repository, to avoid the tedium and inevitable divergences that would arise from having two sources of truth. I don't know what other changes to the separate extension spec would be called for, but probably at least describing how the khr extensions are now found in the spec, and pointing there, would be appropriate - that's in your wheelhouse, I think.

If the WG is ready for that process transition, then nothing besides syncing the mutable extensions with recent changes is really holding up merging. I do need to make some updates to the build to handle the reference pages better. Historically the KHR refpages were kept as separate, static asciidoc files that were merged with the refpage source extracted from the API and C specs. Most of the static files will need to be removed as well, and I may need to make some minor changes to the scripts as I have not looked closely at how it's dealing with the extension appendices yet and they are structured a bit differently than the Vulkan appendices.

I can handle all of that this week, the important thing is that the WG be ready to transition to khr development in the spec context. In addition to ongoing changes to published extensions, any unpublished khr extensions in development should also be done in the context of the API spec going forward. That will probably need some sort of process document for people to refer to, though it's probably not immediately urgent unless you are planning to ratify and release new khr extensions soon.

**Comment by neiltrevett** [comment]
Thanks @oddhack for the above context. We discussed again in today's call - and wish to move forward quickly to merge the new spec.  Can you let us know ASAP when all the needed actions from your side are complete so @bashbaug can merge? Thanks!

**Comment by oddhack** [comment]
> Thanks @oddhack for the above context. We discussed again in today's call - and wish to move forward quickly to merge the new spec. Can you let us know ASAP when all the needed actions from your side are complete so @bashbaug can merge? Thanks!

Yes. I'll clean up remaining issues this week and it should be mergeable before the next CL telecon.

**Comment by oddhack** [comment]
@neiltrevett @bashbaug this is ready now. If possible, I recommend merging #1081 to the branch underlying #950 before merging #950. If the WG is not willing to take that step (removing the separate khr extension documents) yet, it can be retargeted to main and deferred. Any outstanding khr extension work in branches should be re-done as part of the API and C specs ASAP, to avoid divergences and rebasing pain.

I am pretty confident in how this looks now, but definitely review needed before publishing. I'm sure there will be some minor cleanup to deal with, and it would be an excellent idea to run a link-checker over the C and API specs prior to publication - there are many more internal links and a fair number of external links from the API spec extension appendices into the C spec.

I *strongly* recommend squash-on-merge when accepting #950. There is a huge amount of churn in the commits as this took many months to do incrementally, none of which is of particular interest to downstream extension work.

**Comment by oddhack** [comment]
Another good consistency check will be a careful comparison of 'makeSpec -spec core chtml apihtml' with the current HTML from main branch. I have not done this for a while and while the differences should not be big, there are unavoidable diffs. N.b. I find 'wdiff -3 new.html old.html' to be quite useful when changes are small. There are some HTML comparison tools we use in Vulkan but they are... not great. I wish I knew of a great one.

**Comment by bashbaug** [comment]
Discussed in the March 12th teleconference and agreed to merge these changes in principle.  We'll wait for the CI builds to complete and give one more quick review now that #1081 is merged, but assuming all goes well we'll merge these changes ASAP.

**Comment by KhronosWebservices** [comment]
My bad. Used the wrong account for something which ate up the API rate limit for the account we use for cla-assisstant. :(

 Will be up and running again within the hour.

**Comment by oddhack** [comment]
@bashbaug if you want, I can just go ahead and merge this - not sure if you have those permissions on the repo or not? You probably should.

Since I'm a Khronos member I'm already covered for my contributions.

**Comment by KhronosWebservices** [comment]
CLA-Assistant is responding again. Might be good to close and re-open this PR, have the reviewer sign-off, and the CLA check should then be available.

**Review by bashbaug — APPROVED** [review]
Merging as discussed in the March 12th and March 19th teleconferences.

**Comment by SunSerega** [comment]
Can someone please explain what exactly being `ratified` means here? Is it something similar to being an ARB extension in OpenGL? Does it mean an extension is planned to be added to the future version of the core spec?

**Comment by oddhack** [comment]
> Can someone please explain what exactly being `ratified` means here? Is it something similar to being an ARB extension in OpenGL? Does it mean an extension is planned to be added to the future version of the core spec?

Yes, it is similar. Ratification has no effect on developers or users, but is of concern to implementers who want to be sure they do not implement functionality that is not covered by the reciprocal IP licensing obligations of the Khronos IP agreements. It means that Khronos has agreed that a core version or extension is covered by those agreements. There is no obligation or expectation that a ratified extension will become part of a future core API version, and many will not. Conversely non-ratified vendor or multivendor extensions are sometimes promoted to the core API, at which time they are ratified.

**Comment by SunSerega** [comment]
Thanks. Also, in `registry.rnc`:
```
#   depends - boolean expression of API and/or extension names
#       upon which this extension depends.
```
I would intuitively expect `and`/`&` as operators since this value is said to be a boolean expression, but currently I only see a `+` operator.
Well, it can be seen as a boolean operator too, but how about specifying this in the `registry.rnc` in more detail?
And does this mean the value can be something like `(a+b) * (c+d)` in the future?

And another thing:
```
        <extension name="cl_arm_get_core_id" condition="defined(CL_VERSION_1_2)" supported="opencl">
        <extension name="cl_ext_image_requirements_info" condition="defined(CL_VERSION_1_2)" supported="opencl">
        <extension name="cl_ext_image_from_buffer" condition="defined(CL_VERSION_1_2)" supported="opencl">
```
These were probably supposed to use this `depends` attribute instead of `condition`, since it now supports both extension and feature dependencies?

**Pull request merged**
